java/WordCountOptimized.java

```diff
@@ -217,33 +207,66 @@ public static int tokenizeAndSubmitBlock(TObjectIntCustomHashMap<byte[]> counts,
         }
         return 0;
     }
-    
-    private static void submitWord(TObjectIntCustomHashMap<byte[]> m, byte[] word) {
-        m.adjustOrPutValue(word, 1, 1);        
+
+    private static final int TABLE_SIZE = 1024 * 1024;
+    private static final int TABLE_MASK = TABLE_SIZE - 1;
+
+    static class WordMap {
+        final byte[][] keys = new byte[TABLE_SIZE][];
+        final int[] values = new int[TABLE_SIZE];
+        int size = 0;
     }
-    
+
+    private static int submitWord(byte[][] keys, int[] values, int size,
+                                   byte[] word, int off, int len) {
+        int hashCode = (int) XX_HASH.hashBytes(word, off, len);
+        int index = hashCode & TABLE_MASK;
+        while (true) {
+            byte[] key = keys[index];
+            if (key != null) {
+                if (equal(key, word, off, len)) {
+                    values[index]++;
+                    return size;
+                }
+                index = (index + 1) & TABLE_MASK;
+            } else {
+                keys[index] = Arrays.copyOfRange(word, off, off + len);
+                values[index] = 1;
+                return size + 1;
+            }
+        }
+    }
+
+    private static boolean equal(byte[] key, byte[] word, int off, int len) {
```

Comment:
of cause it should be tested (that I cannot do), but my bet is that it won't be faster. Especially if JVM recognizes byte comparison pattern and replaces it with `memcmp`.

Note how `java.lang.String` doesn't compare hash codes, even given them pre-computed. But in this case storing hash codes requires additional storage and penalty (that String have already paid). OTOH, `String` should handle hashcode = 0 (not computed) case.